synapse/rest/client/v2_alpha/capabilities.py

```diff
@@ -27,21 +33,17 @@ class CapabilitiesRestServlet(RestServlet):
 
     PATTERNS = client_patterns("/capabilities$")
 
-    def __init__(self, hs):
-        """
-        Args:
-            hs (synapse.server.HomeServer): server
-        """
+    def __init__(self, hs: "HomeServer"):
         super().__init__()
         self.hs = hs
         self.config = hs.config
         self.auth = hs.get_auth()
         self.store = hs.get_datastore()
 
-    async def on_GET(self, request):
+    async def on_GET(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
         requester = await self.auth.get_user_by_req(request, allow_guest=True)
-        user = await self.store.get_user_by_id(requester.user.to_string())
-        change_password = bool(user["password_hash"])
+        await self.store.get_user_by_id(requester.user.to_string())
+        change_password = self.config.password_enabled
```

Comment:
(Similar to the comment I left on https://github.com/matrix-org/synapse/pull/9587/files#r592377951)

I'm not sure this is the right way to do this, you can have both SSO user and password users in the same server. Does it make sense to allow a user who signs in via SSO to set a password? I'm not sure.

Reply:
I had the same question in top of this PR.
Do you know if Synapse has an function to get information of login type of an user?

Reply:
There's a `get_identity_providers_for_user` method on `SsoHandler` which would return an empty dict if there's no valid IdPs to log in with. I'm just unsure if this is the proper thing to do or not. 😢

Reply:
Ok.
I think about: Is it possible eg. login in web app with SSO and in mobile app with Synapse local db password? Can an user use SSO and password login, can it switch choose how to login?

Reply:
I have found `_get_available_ui_auth_types` and did an update.
https://github.com/matrix-org/synapse/blob/33a02f0f52a5cd793c0d123463d8a7a1e3f4f198/synapse/handlers/auth.py#L387-L388

Reply:
From the conversation we had in #synapse-dev:matrix.org, it seems that we actually want this to be based on is whether passwords are enabled or not (i.e. we *do* want SSO users to be able to set a password if passwords are enabled).

The set password handler only allows setting this if the local password DB is available:

https://github.com/matrix-org/synapse/blob/3101a304fc2093c2eb23e38ce2b11a853a2e0c9b/synapse/handlers/set_password.py#L47-L48

It also doesn't really make sense if passwords are completely disabled, so I think we want to keep this on:

`change_password = self.config.password_enabled and self.config.password_localdb_enabled`

This should be set in the `__init__` since it is the same for all users.

Note that this logic was originally added in #4472. That doesn't seem to mention why it is checking for the password hash though.